### DISCUSSION TITLE
Excessive Client CPU Usage

### DISCUSSION THREAD

**Discussion by kwalker99** [apohl79/audiogridder#1000] [Help]
So we have been testing AG in the studio and have come up against a serious issue where multiple AG client side plugins and sometimes just one, pegs an entire core at 100% or very high usage all the time. We also note that it gets worse with smaller AG block sizes.

Client is a Macbook Pro, Intel i9 8 core, 32GB ram, Big Sur and Logic X, latest version etc, UAD Apollo 8, 128 - 512 samples
Server is a Lenovo thinkstation P350 SFF, with a Intel i9 8 core cpu, 32GB ram, windows 10, core parking disabled, High Performance Power Profile and windows de-bloated etc
Client/Server are connected at 1Gig with copper and via a switch etc.

After doing some extensive testing we have concluded that Logic is unable to fully load balance AG instances across cores, like it does for other plugins or there is some sort of single process that is in the background moving data. 

Is this by design?

Can anything be done to improve this behavour allowing the DAW to better load balance the AG client plugins?

As you can see from the pictures, there is some additional load balanced cpu usage, but a single core is heavily loaded, regularly hitting 100% and causing logic to stop or the audio stops working for all plugins.

![299300192_404099051836611_6656542873580442935_n](https://user-images.githubusercontent.com/18462107/186565438-352d0dff-adac-4d03-ae5a-543d39dd5f97.png)
![300604189_379756404348749_8078285893933876090_n](https://user-images.githubusercontent.com/18462107/186565450-03861fd9-96b0-4773-9d5e-1d8d1dbbea70.png)

**Comment by apiatto** [comment]
My set up is similar to yours and It happens to me too randomly with beta 12. I had also another issue: sometimes when I change the buffer size (fixed for all the plugins) all the tracks sound bad.

**Comment by apohl79** [comment]
The "load" (the higher the latency, the higher the meter) on the last core in logic is I/O related. I can't really see details on your screenshots and I don't have access to UAD hardware right now, but I guess you have a UAD plugin loaded before AG or an apple I/O plugin, or it might behaving like this in general with an apollo, not sure. This last core is spiking much faster than the others, if AG blocks too long (likely because of much stricter realtime requirements interfering with hardware). 

Are you testing without buffering in AG? Have you tried 1 or two blocks?

Are you testing with remote plugins loaded? If you have plugins loaded, it might be one of them causing a spike.

AG is logging a few performance measurements and statistics. So you can check the plugin and server log files. General audio stats look like this:

[11:38:56.682|stats|7f929dc9f1e0] audio: total 9675, rps 187.88, 95th 3.30612ms, avg 2.06ms, min 1.20ms, max 15.92ms
[11:38:56.682|stats|7f929dc9f1e0] audio:  dist 0-2ms 92.58%, 2-4ms 6.99%, 4-6ms 0.37%, 6-8ms 0.04%, 8-10ms 0.00%, 10-12ms 0.00%, 12-14ms 0.01%, 14-16ms 0.01%, 16-18ms 0.00%, 18-20ms 0.00%, >20ms 0.00%

This is a plugin message and it says that 92.58% of the audio precessing calls in the last minute took between 0-2ms, in average 2.06ms and the 95th percentile took 3.3ms etc. but there was at least an outlier which took >12ms and another one >14ms.

If the AG plugin is waiting longer than 15ms for data, it will print a trace message, that shows what parts of the processing flow took how long (usually it would spend most time waiting for data in a read call, see below).

[11:38:56.180|client|7f92de73ebe0] process block took 16.1223ms (_5235c8d9-64eb-4334-9ff7-9c51361f239e_)
[11:38:56.181|client|7f92de73ebe0]   - pb_bypass_chk: 0.014451ms
[11:38:56.181|client|7f92de73ebe0]   - pb_prep: 0.053011ms
[11:38:56.181|client|7f92de73ebe0]   - pb_get_streamer: 0.002544ms
[11:38:56.181|client|7f92de73ebe0]   - pb_ch_map: 0.015996ms
[11:38:56.181|client|7f92de73ebe0]     + as_prep: 0.034738ms
[11:38:56.181|client|7f92de73ebe0]     + as_send: 0.353951ms
[11:38:56.181|client|7f92de73ebe0]   = pb_send: 0.397069ms
**[11:38:56.181|client|7f92de73ebe0]     + as_read: 15.6108ms** <- thats the audiostreamer reading from the network
[11:38:56.181|client|7f92de73ebe0]   = pb_read: 15.6241ms
[11:38:56.181|client|7f92de73ebe0]   - pb_ch_map_reverse: 0.012946ms
[11:38:56.181|client|7f92de73ebe0]   - pb_finish: 0.001429ms

The server is printing a similar message if it needs more than 10ms to process an audio block. If you check the server log file, you might find a related message (see the uuid in the first line):

[11:38:56.161|worker|7f98f0f7a778] process audio took 14.6584ms (_5235c8d9-64eb-4334-9ff7-9c51361f239e_)
[11:38:56.161|worker|7f98f0f7a778]   - aw_lock: 0.001713ms
[11:38:56.161|worker|7f98f0f7a778]   - aw_prep: 0.001619ms
[11:38:56.161|worker|7f98f0f7a778]     + chain_lock: 0.004867ms
[11:38:56.161|worker|7f98f0f7a778]       + proc_loaded_ok: 0.003065ms
[11:38:56.161|worker|7f98f0f7a778]       + proc_got_backend: 0.001059ms
[11:38:56.161|worker|7f98f0f7a778]       + proc_process_0: 14.3781ms
**[11:38:56.161|worker|7f98f0f7a778]     = chain_process: VIRIDIAN2: 14.391ms**
[11:38:56.161|worker|7f98f0f7a778]       + proc_loaded_ok: 0.00224ms
[11:38:56.161|worker|7f98f0f7a778]       + proc_got_backend: 0.000892ms
[11:38:56.161|worker|7f98f0f7a778]       + proc_process_0: 0.247871ms
[11:38:56.161|worker|7f98f0f7a778]     = chain_process: Ozone 9: 0.255415ms
[11:38:56.161|worker|7f98f0f7a778]   = aw_process: 14.6529ms

When I got the I/O meter in logic (last core) to spike at 100% with no AG buffering in my tests, I got this messages printed. In my case you see that VIRIDIAN2 took close to >14ms.

You should also make sure your network is fast. A ping between my server and daw computer is stable between 0.3-0.7ms.

**Reply by kwalker99** [reply]
The network is sub 1ms, as it will be for most of the people using 1gig ethernet with almost nothing on the network or point to point

No UAD plugins are being used, just a few native software synths client side and the TC Nonlin2 reverb server side, for each AG plugin instance. 

It's not a spike, it's a constant overhead, that increases as more AG plugins are enabled.

Sadly your code has no comments so is hard to follow, do you have a architectural diagram? 

How are the plugins getting data out of the daw and into the network? I see you are using UNIX sockets and have a ring buffer, is that per plugin instance or is a single ring buffer being used for all plugin instances? 

By running multiple AG servers on one remote machine, does this change the buffering behavior client side?

It is really hard to understand how this has been designed (by the way, I am a Solutions Architect so work on software integration projects by day), I would really like to help and have some interesting history we should talk about offline.

https://user-images.githubusercontent.com/18462107/186781914-ccad3f1b-33a1-4ce4-a2a6-5580a3ed29e5.png
https://user-images.githubusercontent.com/18462107/186782009-f0dfb08e-909f-48bc-b227-98bab47142b5.png

**Reply by apohl79** [reply]
I see. I did a bit more testing and logic seems to behave similarly when using instruments. Once I load an instrument like alchemy, I see the last meter rising. The smaller the audio interface I/O buffer the higher the meter. If you add AG without buffering, each audio processing call will have to block to do network I/O. So the smaller your interface buffer, the more likely you will need AG buffering. 

(update): I found this thread: https://www.logicprohelp.com/forums/topic/133646-understanding-the-performance-meter/

So once you add an empty audio track and select it, the "live thread" (which should be the last meter) drops to 0. So this explains the behaviour.

> How are the plugins getting data out of the daw and into the network? I see you are using UNIX sockets and have a ring buffer, is that per plugin instance or is a single ring buffer being used for all plugin instances?

This is implemented in the AudioStreamer class (send & read), which is managed by the Client class, which manages all server connections. The audio/midi entry point in the plugin is the processBlock method of the PluginProcessor class. Unix sockets are only used if you are running the server locally (on the same machine as you run your daw) as they are an IPC mechanism only.

The ring buffer is used to delay the sample stream if the plugin is bypassed (see processBlockBypassed). The audio streamer works with buffer queues, if you enable buffering in AG. Without buffering the audio sample buffer passed by the daw is serialized directly to the socket and filled back reading from the socket.

> By running multiple AG servers on one remote machine, does this change the buffering behaviour client side?

no

> I would really like to help and have some interesting history we should talk about offline.

Sure. Any help is very much welcome! You can mail me at audiogridder at e47.org.

**Comment by nodevnul** [comment]
Hi.

Check in

Logic --> Preferences --> Audio --> Multithreading

You can choose between

Playback tracks
Playback and Live Input

Check also the other settings to optimize for your usecase.

Now, Logic by design will dedicate a full CPU core to live input processing in order to provide the capability to play virtual instruments with low latency buffer settings. That explains the Logic CPU meter behavior. At least Apple made this function much more efficient from LP 10.4 to 10.5 and onwards. Hence there is nothing AG can do here I believe. Yet usually this should not be of a problem.

Regards

**Comment by kwalker99** [comment]
Yep, we have tried this already. AG is still being treated strangely by
Logic, other hosts like Bitwig or Cubase don't have this issue. It's not
related to Live tracks btw, even when selecting just a midi track, 1 core
gets hammered by AG in Logic.

For mixing larger buffer sizes are very usable, but for running instruments
and other plugs with a desire for low latency, then using some alternative
audio I/O seems to be the only option.

**Reply by nodevnul** [reply]
Ok.

Have you tried Vienna Ensemble Pro 7 ? I think there is a demo option to test it. It is my understanding that VEP has been a trusted and well working Client-Server Plugin host for composers, also in Logic.

Therefore I rather would say that AG has a problem here. There are also other Client-Server-data-Transmission-solutions that obviously work in Logic (like Softube Console 1, SSL analoge console remote control, Audiomovers, Sonobus…).

I have VEP 7 and an classic MP5.1 and a M1Max yet are preparing to put all together. Maybe I can report sometime what the state with this setup is.

**Comment by ZeChArtiahSaher** [comment]
I agree, in ableton I get the same issue as well that CPU skyrockets if there is some network latency somewhere. Something should be converted to async or partial processing I think. I suspect busy waiting happens somewhere

**Reply by nodevnul** [reply]
Yes, Network is something to check for shure.

The audio hosting / streaming should not go through your standard routers / network you use for www, NAS and alike. It should have it's dedicated NICs and router, no USB, dedicated Ethernet PCIe / TB and alike.

I have setup Dante like that in the studio and it runs exactly zero errors like a tank.
